src/js/utils/DSLUpdateUtil.js

```diff
@@ -0,0 +1,197 @@
+   *
+   * @param {DSLExpression} expression - The expression to update
+   * @param {Array} nodes - The node(s) to append
+   * @returns {DSLExpression} expression - The updated expression
+   */
+  applyAdd(expression, nodes) {
+    let expressionValue = expression.value;
+    const appendNodes = nodes.map((node) => {
+      return DSLUtil.getNodeString(node);
+    });
+
+    // Add space only if we have an expression already
+    if (expressionValue) {
+      expressionValue += ' ';
+    }
+
+    // Return new expression
+    return new DSLExpression(
+      expressionValue + appendNodes.join(' ')
+    );
+  },
+
+  /**
+   * Update the given nodes in the expression with the new array of nodes
+   * taking correcting actions if the lengths do not match.
+   *
+   * @param {DSLExpression} expression - The expression to update
+   * @param {Array} nodes - The node(s) to update
+   * @param {Array} newNodes - The node(s) to update with
+   * @returns {DSLExpression} expression - The updated expression
+   */
+  applyUpdate(expression, nodes, newNodes) {
+    const updateCount = Math.min(nodes.length, newNodes.length);
+    let expressionValue = expression.value;
+    let offset = 0;
+
+    // First update nodes
+    for (let i=0; i<updateCount; ++i) {
+      const updateNode = nodes[i];
+      const withNode = newNodes[i];
+
+      // Update expression value
+      const newValue = DSLUpdateUtil.updateNodeString(
+        expressionValue, updateNode, withNode, offset
+      );
+
+      // This action may have shifted the location of the tokens
+      // in the original expression. Update offset.
+      offset += newValue.length - expressionValue.length;
+
+      expressionValue = newValue;
+    }
+
+    // Then delete nodes
+    if (newNodes.length < nodes.length) {
+      // We are deleting from the last to the first in order to keep
+      // the offsets intact so we don't have to recalculate
+      for (let i=nodes.length-1; i>=updateCount; --i) {
+        const deleteNode = nodes[i];
+        expressionValue = DSLUpdateUtil.deleteNodeString(
+          expressionValue, deleteNode, offset
+        );
+      }
+    }
+
+    // And finally insert new nodes
+    if (newNodes.length > nodes.length) {
+      const appendNodes = newNodes.slice(updateCount).map((node) => {
+        return DSLUtil.getNodeString(node);
+      });
+
+      // Add space only if we have an expression already
+      if (expressionValue) {
+        expressionValue += ' ';
+      }
+
+      expressionValue += appendNodes.join(' ');
+    }
+
+    // Compile and return the new expression
+    return new DSLExpression(expressionValue);
+  },
+
+  /**
+   * Delete the given list of nodes from the expression
+   *
+   * @param {DSLExpression} expression - The expression to update
+   * @param {Array} nodes - The node(s) to delete
+   * @returns {DSLExpression} expression - The updated expression
+   */
+  applyDelete(expression, nodes) {
+    var offset = 0;
```

Comment:
`let` is supposed to be block-scoped, meaning that it won't be visible to the function inside (even though babel just converts it into `var`) in this case `var` has a reason for being there.

But yeah, I will use reduce with `{expression, offset}` instead.